qiskit/pulse/schedule.py

```diff
@@ -171,77 +179,135 @@ def _instructions(self, time: int = 0):
                 :class:`~qiskit.pulse.Instruction`
                 starts at and the flattened :class:`~qiskit.pulse.Instruction` s.
         """
-        for insert_time, child_sched in self._children:
+        for insert_time, child_sched in self.children:
             yield from child_sched._instructions(time + insert_time)
 
-    def union(self, *schedules: Union[ScheduleComponent, Tuple[int, ScheduleComponent]],
-              name: Optional[str] = None) -> 'Schedule':
-        """Return a new schedule which is the union of both ``self`` and ``schedules``.
+    def union(self,
+              *schedules: Union[ScheduleComponent, Tuple[int, ScheduleComponent]],
+              name: Optional[str] = None,
+              mutate: bool = False
+              ) -> 'Schedule':
+        """Return a schedule which is the union of both ``self`` and ``schedules``.
 
         Args:
-            *schedules: Schedules to be take the union with this ``Schedule``.
+            schedules: Schedules to be take the union with this ``Schedule``.
             name: Name of the new schedule. Defaults to the name of self.
+            mutate: Perform operation by mutating this schedule. Otherwise return
+                a new ``Schedule``.
         """
         warnings.warn("The union method is deprecated. Use insert with start_time=0.",
                       DeprecationWarning)
+        if mutate:
+            return self._mutable_insert(0, *schedules)
+        else:
+            return self._immutable_insert(0, *schedules, name=name)
+
+    # pylint: disable=arguments-differ
+    def shift(self,
+              time: int,
+              name: Optional[str] = None,
+              mutate: bool = False
+              ) -> 'Schedule':
+        """Return a schedule shifted forward by ``time``.
+
+        Args:
+            time: Time to shift by.
+            name: Name of the new schedule. Defaults to the name of self.
+            mutate: Perform operation by mutating this schedule. Otherwise return
+                a new ``Schedule``.
+        """
+        if mutate:
```

Comment:
This is a pretty big change to the API and the structure of all these methods. I don't remember seeing this in the RFC. What alternative approaches did you attempt, what went wrong with those, and what are the pros and cons of this approach?

Reply:
This or a separate mutable insert is the only way to build up the pulse builder interface to maintain a handle on the schedule being built. We must mutate the exposed schedule given the syntax I show.

Reply:
Have we considered making schedules mutable by default?

Reply:
I would like to consider moving towards this very much! Unfortunately, it may break any code that calls `insert` and `append` directly. However, we can still maintain the `+/|/<<` syntax quite easily. We should consider a transition path separately.